pkg/virtctl/convertmachinetype/massmachinetypetransition/update_machine_type.go

```diff
@@ -0,0 +1,134 @@
+package massmachinetypetransition
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	k8sv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	k6tv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+)
+
+// using these as consts allows us to easily modify the program to update as newer versions are released
+// we generally want to be updating the machine types to the most recent version
+const (
+	LatestMachineTypeVersion           = "rhel9.2.0"
+	MinimumSupportedMachineTypeVersion = "rhel9.0.0"
+)
+
+var (
+	VmisPendingUpdate = make(map[string]struct{})
+	exitJob           = make(chan struct{})
+
+	// by default, update machine type across all namespaces
+	namespace     = k8sv1.NamespaceAll
+	labelSelector = ""
+
+	// by default, should require manual restarting of VMIs
+	restartNow = false
+)
+
+func VerifyMachineType(machineType string) (bool, string) {
+	// if the machine type is "q35", we want the VM to behave as though its machine type has been updated
+	if machineType == "q35" {
+		return true, machineType
+	}
+
+	machineTypeSubstrings := strings.Split(machineType, "-")
+
+	if len(machineTypeSubstrings) == 3 {
+		version := machineTypeSubstrings[2]
+		if strings.Contains(version, "rhel") && version < MinimumSupportedMachineTypeVersion {
+			machineTypeSubstrings[2] = LatestMachineTypeVersion
+			machineType = strings.Join(machineTypeSubstrings, "-")
+			return true, machineType
+		}
+	}
+	return false, machineType
+}
+
+func UpdateMachineTypes(virtCli kubecli.KubevirtClient) error {
+	listOpts := &k8sv1.ListOptions{}
+	if labelSelector != "" {
+		listOpts = &k8sv1.ListOptions{
+			LabelSelector: labelSelector,
+		}
+	}
+	vmList, err := virtCli.VirtualMachine(namespace).List(context.Background(), listOpts)
+
+	if err != nil {
+		return err
+	}
+
+	for _, vm := range vmList.Items {
+		machineType := vm.Spec.Template.Spec.Domain.Machine.Type
```

Comment:
The machine type is only in the VMI Status, not VM Status. So to check all of the VMs it is correct to use the VM Spec to determine the machine type (otherwise we would end up skipping over non-running VMs entirely).

Reply:
As it is implemented now, I'm not sure how the code distinguishes between these cases:
1. a VM that's running, has `q35` in spec, but in practice uses an outdated machine type.
2. a VM that's running, has `q35` in spec, and runs on the latest machine type.

The one needs to restart, the other doesn't.
I think that eventually there's a need to fetch the actual version from the VMI status for running VMs. If the VM is not running and has `q35` in spec, we can consider it as already updated (assuming the machine type is updated automatically in that case, which I _think_ is correct).

Reply:
Doesn't this [already happen](https://github.com/kubevirt/kubevirt/blob/fd36738c1be150e67acf9d9235db1261136bdcbb/pkg/virtctl/convertmachinetype/massmachinetypetransition/update_machine_type.go#L94)?  After it is determined if the VM is running it checks if the VMI actually has an outdated machine type before attempting to restart/add label. Otherwise it just skips it and moves to the next VM.

Reply:
I've also verified that restarting the VM with `q35` in spec and outdated machine type in VMI status updates the machine type automatically on reboot.

Reply:
>Doesn't this [already happen](https://github.com/kubevirt/kubevirt/blob/fd36738c1be150e67acf9d9235db1261136bdcbb/pkg/virtctl/convertmachinetype/massmachinetypetransition/update_machine_type.go#L94)? After it is determined if the VM is running it checks if the VMI actually has an outdated machine type before attempting to restart/add label. Otherwise it just skips it and moves to the next VM.

Looks good! :)